binding.c

```diff
@@ -227,6 +264,309 @@ bare_dns_lookup(js_env_t *env, js_callback_info_t *info) {
+  intrusive_list_init(&resolver->tasks);
+
+  struct ares_options opts;
+  opts.sock_state_cb = bare_dns__on_socket_change;
+  opts.sock_state_cb_data = resolver;
+
+  err = ares_init_options(&resolver->channel, &opts, ARES_OPT_SOCK_STATE_CB);
+
+  if (err != ARES_SUCCESS) {
+    js_throw_error(env, NULL, ares_strerror(err));
+    return NULL;
+  }
+
+  resolver->env = env;
+
+  err = js_add_deferred_teardown_callback(env, bare_dns__on_resolver_teardown, (void *) resolver, &resolver->teardown);
+  assert(err == 0);
+
+  return handle;
+}
+
+void
+bare_dns__on_resolve_txt(void *data, ares_status_t status, size_t timeouts, const ares_dns_record_t *dnsrec) {
+  int err;
+
+  bare_dns_resolve_req_t *req = (bare_dns_resolve_req_t *) data;
+
+  js_env_t *env = req->env;
+
+  js_handle_scope_t *scope;
+  err = js_open_handle_scope(env, &scope);
+  assert(err == 0);
+
+  js_value_t *ctx;
+  err = js_get_reference_value(env, req->ctx, &ctx);
+  assert(err == 0);
+
+  js_value_t *cb;
+  err = js_get_reference_value(env, req->cb, &cb);
+  assert(err == 0);
+
+  err = js_delete_reference(env, req->cb);
+  assert(err == 0);
+
+  err = js_delete_reference(env, req->ctx);
+  assert(err == 0);
+
+  js_value_t *args[2];
+
+  if (status == ARES_SUCCESS) {
+    unsigned char *buf;
+    size_t len = 0;
+
+    status = ares_dns_write(dnsrec, &buf, &len);
+    assert(status == ARES_SUCCESS);
+
+    struct ares_txt_ext *reply;
+
+    status = ares_parse_txt_reply_ext(buf, len, &reply);
+    assert(status == ARES_SUCCESS);
+
+    err = js_get_null(env, &args[0]);
+    assert(err == 0);
+
+    js_value_t *result;
+    err = js_create_array(env, &result);
+    assert(err == 0);
+
+    uint32_t i = 0;
+
+    for (struct ares_txt_ext *next = reply; next != NULL; next = next->next) {
+      js_value_t *chunk;
+      err = js_create_string_utf8(env, (utf8_t *) next->txt, next->length, &chunk);
+      assert(err == 0);
+
+      if (next->record_start) {
+        js_value_t *record;
+        err = js_create_array(env, &record);
+        assert(err == 0);
+
+        err = js_set_element(env, record, 0, chunk);
+        assert(err == 0);
+
+        err = js_set_element(env, result, i++, record);
+        assert(err == 0);
+      } else {
+        js_value_t *record;
+        err = js_get_element(env, result, i, &record);
+        assert(err == 0);
+
+        uint32_t len;
+        err = js_get_array_length(env, record, &len);
+        assert(err == 0);
+
+        err = js_set_element(env, record, len, chunk);
+        assert(err == 0);
+      }
+    }
+
+    args[1] = result;
+
+    ares_free_data(reply);
+  } else {
+    js_value_t *message;
+    err = js_create_string_utf8(env, (utf8_t *) ares_strerror(status), -1, &message);
+    assert(err == 0);
+
+    err = js_create_error(env, NULL, message, &args[0]);
+    assert(err == 0);
+
+    err = js_get_null(env, &args[1]);
+    assert(err == 0);
+  }
+
+  if (!req->resolver->exiting) js_call_function(env, ctx, cb, 2, args, NULL);
```

Comment:
```suggestion
  js_call_function(env, ctx, cb, 2, args, NULL);
```

The code must abort much earlier on teardown so we don't inadvertently end up calling into JavaScript.